app/src/utils/rateLimits.ts

```diff
@@ -0,0 +1,32 @@
+import { sql } from "kysely";
+import { kysely } from "~/server/db";
+
+type RateLimitConfig = {
+  capacity: number; // Maximum tokens in the bucket
+  fillRate: number; // Tokens added to the bucket per second
+};
+
+export const rateLimit = async (key: string, config: RateLimitConfig): Promise<boolean> => {
+  const tokensToAdd = sql`((extract(epoch from now()) - extract(epoch from "RateLimit"."lastUpdated")) * ${config.fillRate})`;
+
+  const currentTokenCount = sql`least(${config.capacity}, "RateLimit".tokens + ${tokensToAdd})`;
+
+  const query = sql<{ allowed: boolean; tokensLeft: number }>`
+    insert into "RateLimit"(key, tokens, allowed, "lastUpdated")
+    values (${key}, ${config.capacity} - 1, true, now())
+    on conflict (key) do update set
+      tokens = greatest(0, ${currentTokenCount} - 1),
+      allowed = round(${currentTokenCount}) > 0,
+      "lastUpdated" = now()
```

Comment:
By this logic, if we get hammered with requests, won't we keep updating "lastUpdated" while not actually sending any requests at all?

We should avoid setting "lastUpdated" if currentTokenCount is 0.